[distributed] Require explicit assignment of `actorSystem`

kavon opened the pull request:
As discussed in https://forums.swift.org/t/specifying-the-controlling-actor-system-with-distributed-actors-se-0336-0344/55584

As of now, we require exactly one parameter of a designated initializer of a distributed actor to conform to DistributedActorSystem. Then we implicitly initialize self.actorSystem and self.id based using that parameter. The problem is that there is no explicit use of this parameter, so it seems / feels visually like it is not needed.

Instead, this PR changes things so that we require users to manually initialize the `actorSystem` property in the designated initializer's body, dropping the parameter requirement. Since we need to right-away initialize the ID after doing this, we would keep the implicit initialization of the ID.

Resolves rdar://89582844 and rdar://86543336

kavon committed:
remove restrictions on dist-actor init parameters

we will no longer require exactly one ActorSystem-conforming parameter.

kavon committed:
rename default initializer's ActorSystem parameter name to actorSystem

Konrad, myself, and others feel the longer name would be better, rather than just "system"

kavon committed:
remove dependence on system parameter in SILGen/DefiniteInit

ktoso commented on include/swift/SILOptimizer/Utils/DistributedActor.h:
```diff
@@ -32,10 +32,6 @@ class SILLocation;
 
-/// Finds the first `DistributedActorSystem`-compatible parameter of the given function.
-/// \returns nullptr if the function does not have such a parameter.
-SILArgument *findFirstDistributedActorSystemArg(SILFunction &F);
```
👍

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -235,7 +276,8 @@ static void emitIDInit(SILGenFunction &SGF, ConstructorDecl *ctor,
 
-  SILValue actorSystem = findFirstDistributedActorSystemArg(F);
+  // since we're doing this only for the implicitly defined ctors, grab from arg
+  SILValue actorSystem = findFirstDistributedActorSystemArg(SGF.F);
```
👍

ktoso commented on lib/Sema/CodeSynthesis.cpp:
```diff
@@ -301,8 +301,8 @@ static ConstructorDecl *createImplicitConstructor(NominalTypeDecl *decl,
-      // Create the parameter.
-      auto *arg = new (ctx) ParamDecl(SourceLoc(), Loc, ctx.Id_system, Loc,
+      // Create the parameter. API name is actorSystem, local name is system
+      auto *arg = new (ctx) ParamDecl(SourceLoc(), Loc, ctx.Id_actorSystem, Loc,
```
nice, thanks 👍

kavon commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -787,6 +787,42 @@ static bool isDistributedActorCtor(SILFunction &F) {
   return false;
 }
 
+//static SILValue emitActorPropertyReference(
+//    SILGenFunction &SGF, SILLocation loc, SILValue actorSelf,
+//    VarDecl *property) {
+//  Type formalType = SGF.F.mapTypeIntoContext(property->getInterfaceType());
+//  SILType loweredType = SGF.getLoweredType(formalType).getAddressType();
+//  return SGF.B.createRefElementAddr(loc, actorSelf, property, loweredType);
+//}
```
delet this

kavon committed:
adding the @_compilerInitialized attribute

This attribute is designed for let-bound variables whose initializing
assignment is synthesized by the compiler. This assignment is
expected to happen at some point before DefiniteInitialization has
run, which is the pass that verifies whether the compiler truly
initialized the variable.

I generally expect that this will never be a user-facing feature, and
that the synthesized assignment happens in SILGen.

kavon committed:
basic implementation of performing an init of the id after an assign of the actorSystem.

The resignID call within the initializer moved into DI, because an assignment to
the actorSystem, and thus initialization of the id, is no longer guaranteed to happen.
Thus, while we previously could model the resignation as a clean-up emitted on all
unwind paths in an initializer, now it's conditional, based on whether the id was
initialized. This is exactly what DI is designed to do, so we inject the resignation
call just before we call the identity's deinit.

ktoso commented on lib/SILGen/SILGenLValue.cpp:
```diff
@@ -4605,6 +4674,13 @@ void SILGenFunction::emitAssignLValueToLValue(SILLocation loc, LValue &&src,
+  // FIXME: in what situations does this lvalue to lvalue assign happen?
+  if (LLVM_UNLIKELY(DistActorCtorContext) &&
+      referencesVar(**(dest.end()-1), DistActorCtorContext->actorSystem))
+    Cleanups.pushCleanup<DistActorIDInitializer>(*DistActorCtorContext);
```
Is this via cleanup actually right...? I thought cleanups are done at the end of a scope?

(Just trying to understand this better)

kavon replied:
Cleanups should be emitted in-order at the end of the evaluation scope in which they were pushed. The caller of `emitAssignToLValue` sets one up for the evaluation of the assignment, but there is also an ArgumentScope established within this function. I might move this clean-up to be just before that ArgumentScope, but otherwise I think it's OK to conditionally push a clean-up here.

More generally, I went with the clean-up approach (for now) because I wanted to make sure we property handle situations like these:

```
init() throws {
  self.actorSystem = try getActorSystem()
}

init() {
  do {
    self.actorSystem = try getActorSystem()
  } catch {
    self.actorSystem = defaultSystem
  }
}
```

Since clean-ups can tell you whether it's being emitted on the unwind side or not. On the unwind / error path of the `try_apply` of `getActorSystem`, we should _not_ try to assign an identity, since we have no actor system (something new that was previously guaranteed with the old approach!). This _might_ mean we need to tweak how `actorReady` calls get injected for synchronous inits (to avoid it on the unwind path), but I'm not sure yet.

ktoso replied:
Missed to thank here; but thanked offline -- thanks for the explanation, makes sense!

ktoso commented on lib/SILGen/SILGenLValue.cpp:
```diff
@@ -4505,6 +4507,67 @@ void SILGenFunction::emitAssignToLValue(SILLocation loc, RValue &&src,
+    return false;
+
+  return static_cast<RefElementComponent const&>(comp).getField() == var;
+}
```
oh nice, so detecting it wasn't so terrible 👍

kavon commented:
What's left for this PR is:
- suppressing missing assignments to `@_compilerInitialized` properties if DI there is / will be another missing assignment / error in DI that would cause the initializer to be rejected. I think the way to do this is to just delay emitting the error diagnostic and have a check at the end to see if we emitted an error already.
- SILGen, runtime, and DI tests.

kavon commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -158,68 +196,163 @@ static void emitActorSystemInit(SILGenFunction &SGF,
-    : actorDecl(actorDecl), self(self) {
-      assert(actorDecl->isDistributedActor());
-    }
+//namespace {
+///// Cleanup to resign the identity of a distributed actor if an abnormal exit happens.
+//class ResignIdentity : public Cleanup {
+//  ClassDecl *actorDecl;
+//  SILValue self;
+//
+//  // This flag remembers whether we assigned an identity at runtime. If it is
+//  // nullptr, then we can unconditionally resign our identity.
+//  AllocStackInst *flag;
+//public:
+//  /// \param flag nonnull means the resignation is conditional based on this
+//  /// storage, because the identity may not be assigned on all unwind paths.
+//  /// If the storage holds the value "true" then an identity has been assigned,
+//  /// so it must be resigned.
```
delet this too

ktoso commented on lib/SILGen/InitializeIdentity.h:
```diff
@@ -0,0 +1,40 @@
+
+/// A clean-up designed to emit an initialization of a distributed actor's
+/// identity upon successful initialization of the actor's system.
+struct InitializeIdentity : Cleanup {
```
minor: want to name it with something Distributed? we're fairly consistently naming all things/files with Distributed... something, maybe `InitializeDistributedActorIdentity`?

kavon replied:
oh yea good point. will do

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -133,23 +133,61 @@ static void emitDistributedIfRemoteBranch(SILGenFunction &SGF, SILLocation Loc,
+  auto &C = F.getASTContext();
+
+  auto *transportProto = C.getProtocol(KnownProtocolKind::DistributedActorSystem);
+  Type transportTy = transportProto->getDeclaredInterfaceType();
```
nitpick `systemTy`

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -133,23 +133,61 @@ static void emitDistributedIfRemoteBranch(SILGenFunction &SGF, SILLocation Loc,
+  auto *module = F.getModule().getSwiftModule();
+  auto &C = F.getASTContext();
+
+  auto *transportProto = C.getProtocol(KnownProtocolKind::DistributedActorSystem);
```
nitpick "system" transport was ancient wording.

```suggestion
  auto *DAS = C.getDistributedActorSystemDecl();
```

or `systemProto` if you prefer 👍  recently started using `DAS` `DA` for the protocol types.

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -133,23 +133,61 @@ static void emitDistributedIfRemoteBranch(SILGenFunction &SGF, SILLocation Loc,
+  // By construction, automatically generated distributed actor ctors have
+  // exactly one ActorSystem-conforming argument to the constructor,
+  // so we grab the first one from the params.
   VarDecl *var = lookupProperty(classDecl, C.Id_actorSystem);
```
👍

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -158,68 +196,163 @@ static void emitActorSystemInit(SILGenFunction &SGF,
+  // thrown an error. In that case, we cannot initialize the identity,
+  // since there is no actorSystem.
+  if (forUnwind == IsForUnwind)
+    return;
```
Thx for nice explanation

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -244,12 +384,27 @@ void SILGenFunction::emitDistributedActorReady(
-  emitActorReadyCall(B, loc, borrowedSelf.getValue(), transport);
+  // --- load the actor system from the actor instance
+  ManagedValue actorSystem;
+  SGFContext sgfCxt; // TODO: is this right?
```
AFAIK (could be totally wrong on this...) this is probably right -- I've seen it in other places where just an empty one was used like that

kavon replied:
ah ok thanks. I spotted that somewhere too and it seemed to work so 🤷🏼

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -244,12 +384,27 @@ void SILGenFunction::emitDistributedActorReady(
+                                *this, loc, borrowedSelf.getValue(), property);
+    actorSystem = emitLoad(loc, actorSystemRef,
+                           getTypeLowering(loweredType), sgfCxt, IsNotTake);
+  }
```
Not sure if this is enough... does this handle a struct system as well? 🤔

ktoso replied:
Hmm... maybe it's right... I was just remembering my trauma from the SIL thunk impl:

```
      auto systemRef = emitActorPropertyReference(
           *this, loc, selfValue.getValue(),
           lookupProperty(selfTyDecl, ctx.Id_actorSystem));

       auto actorSystemTy = systemRef->getType();

       // FIXME: this is wrong for struct with values, and classes?
       // %17 = load %16 : $*FakeActorSystem // users: %21, %20, %18
       SILValue systemLoaded;
       if (actorSystemTy.isAddressOnly(F)) {
         assert(false && "isAddressOnly");
       } else {
         if (actorSystemTy.isAddress()) {
           systemLoaded = B.createTrivialLoadOr(
               loc, systemRef, LoadOwnershipQualifier::Copy);
         } else {
           assert(false);
         }
       }
```

https://github.com/apple/swift/pull/41616/files#diff-375d2a63e22265971efa52970e47979b7414d7f6dc7149100f7a6f9fb7158e0eL826

kavon replied:
I think because it's using the SILGenFunction's `emitLoad`, it will be smart about whether it's a class or struct and what to do in order to perform the load. That's probably why it seems to be working, but I'll add a test to cover both.

ktoso replied:
Nice, thank you!

ktoso commented on lib/SILOptimizer/Mandatory/DIMemoryUseCollector.cpp:
```diff
@@ -223,6 +224,33 @@ SILType DIMemoryObjectInfo::getElementType(unsigned EltNo) const {
 }
 
+/// During tear-down of a distributed actor, we must invoke its
+/// \p actorSystem.resignIdentity method, passing in the \p id from the
```
Minor:

```suggestion
/// \p actorSystem.resignID method, passing in the \p id from the
```

meh sadly during review we were asked to stick to `ID` consistently :) 

C++ function names good to use Identity, that's fine :)

ktoso commented on lib/SILOptimizer/Mandatory/DIMemoryUseCollector.cpp:
```diff
@@ -223,6 +224,33 @@ SILType DIMemoryObjectInfo::getElementType(unsigned EltNo) const {
+/// Since deinitialization consistently happens in-order, according to the
+/// listing returned by \p NominalTypeDecl::getStoredProperties
+/// it is important the the VarDecl for the \p id is synthesized before
+/// the \p actorSystem so that we get the right ordering in DI and deinits.
```
👍

ktoso commented on lib/SILOptimizer/Mandatory/DIMemoryUseCollector.cpp:
```diff
@@ -665,6 +703,16 @@ static SILValue getAccessedPointer(SILValue Pointer) {
+                                           DIUseKind origKind) {
+  // if the write is _not_ auto-generated, it's a bad store.
+  if (!write->getLoc().isAutoGenerated())
+    return DIUseKind::BadExplicitStore;
```
Awesome!

ktoso reviewed (approved):
LGTM so far! This is looking really great; the impl ended up very clean thanks to the new attr as well.

kavon committed:
suppress notes from DI about @_compilerInitialized when reasonable

Since the only user of `@_compilerInitialized` is distributed
actors, I decided to make DI's diagnostics saying that such a
var was not initialized a second-tier note. If there is some other
var associated with the same Memory that is _not_ compiler initialized
AND is also _not_ initialized, then we suppress the suggestion to initialize
the @_compilerInitialized thing, because something else is problematic. Only
when there are @_compilerInitialized things that are still not initialized,
do we emit a note talking about that.

kavon commented:
puttin out a feeler: @swift-ci please smoke test

ktoso commented:
As we unlock the local init tests this also resolves rdar://86543336

ktoso committed:
update distributed_actor_init_local test

kavon committed:
fix code completion test

kavon committed:
more test updates for actorSystem assignment

kavon committed:
more coverage for actorSystem assign

kavon commented:
@swift-ci please test

kavon commented on lib/SILGen/InitializeDistActorIdentity.h:
```diff
@@ -0,0 +1,40 @@
+
+namespace swift {
+
+class AllocStackInst;
```
not needed

kavon commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -244,12 +319,27 @@ void SILGenFunction::emitDistributedActorReady(
-  emitActorReadyCall(B, loc, borrowedSelf.getValue(), transport);
+  // --- load the actor system from the actor instance
+  ManagedValue actorSystem;
+  SGFContext sgfCxt; // TODO: is this right?
```
could probably remove this comment now

kavon commented on lib/SILGen/SILGenFunction.h:
```diff
@@ -2071,6 +2077,17 @@ class LLVM_LIBRARY_VISIBILITY SILGenFunction
+                                 SILValue selfVal,
+                                 SILValue actorSystemVal);
+
+  void registerDistActorIdentityInit();
```
this is dead code

kavon reviewed (commented):
ok this edition looks good. just spotted some really minor clean-ups

ktoso commented:
Cool, let's clean those up separately and land this :) 
I'll likely have small PRs coming around so may include those..
